File: communicationssatellite.py
from sys import stdin,stdout
from math import sqrt

INF = 10**9 # use one billion to represent infinity
N = int(stdin.readline())
#Prim's variant for dense graphs is used
points = [tuple(map(int,stdin.readline().split())) for _ in range(N)]
A,taken = [(INF,v) for v in range(N)],[False for _ in range(N)]
A[0] = (0,0) #start from first point
taken[0] = True; cost = 0; counter = 1
while counter < N: #O(N)
    lowest,v = INF,0 # scan A to get v where A[v][0] is minimum in A
    for i in range(N): #O(N)
        if A[i][0] < lowest:
            lowest = A[i][0]
            v = i
    #add vertices to taken
    if not taken[v]:
        taken[v] = True
        counter += 1
    if not taken[A[v][1]]:
        taken[A[v][1]] = True
        counter += 1
    cost += A[v][0]
    A[v] = (INF,A[v][1]) #prevent picking the same point
    for j in range(N): #O(N)
        if not taken[j] and A[j][0] > sqrt((points[v][0]-points[j][0])**2+(points[v][1]-points[j][1])**2) - points[v][2] - points[j][2]: #subtract radius of two circles from distance between their two centres
            A[j] = (sqrt((points[v][0]-points[j][0])**2+(points[v][1]-points[j][1])**2) - points[v][2] - points[j][2],v)
stdout.write(f'{cost}\n')

File: communicationssatellite.java
import java.io.*;

class Circle implements Comparable<Circle> {// stores coordinates
    private double _first, _second,_third;
  
    public Circle(double f, double s, double t) {
        _first = f;
        _second = s;
        _third = t;
    }
  
    public int compareTo(Circle o) {
      if (Double.compare(this.x(),o.x()) != 0) return Double.compare(this.x(),o.x());
      else if (Double.compare(this.y(),o.y()) != 0) return Double.compare(this.y(),o.y());
      else return Double.compare(this.radius(),o.radius());
    }
  
    public double x() {return _first;}
  
    public double y() {return _second;}

    public double radius() {return _third;}
}

class DoubleIntPair implements Comparable<DoubleIntPair> {// stores coordinates
    private double _first;
    private int _second;
  
    public DoubleIntPair(double f, int s) {
      _first = f;
      _second = s;
    }
  
    public int compareTo(DoubleIntPair o) {
      if (Double.compare(this.first(),o.first()) != 0) return Double.compare(this.first(),o.first());
      else return this.second() - o.second();
    }
  
    public double first() {return _first;}
  
    public int second() {return _second;}
}

public class communicationssatellite {
    public static double distance(double x1, double y1, double x2, double y2){
        return Math.sqrt(Math.pow(x1-x2,2) + Math.pow(y1-y2, 2)); //calculate Euclidean distance
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        double INF = 1000000000; // use 1 billion to represent infinity
        int n = Integer.parseInt(br.readLine());
        Circle[] points = new Circle[n]; // use Prim's variant for dense graphs with O(n^2) time complexity
        DoubleIntPair[] A = new DoubleIntPair[n];
        for (int j = 0; j < n; j++) {
            String[] line = br.readLine().split(" ");
            points[j] = new Circle(Double.parseDouble(line[0]), Double.parseDouble(line[1]),Double.parseDouble(line[2]));
            A[j] = new DoubleIntPair(INF, j); // initialisation
        }
        boolean[] taken = new boolean[n];
        taken[0] = true; A[0] = new DoubleIntPair(0, 0);
        double cost = 0; int counter = 1; // cost keeps track of total edge weights in MST, while counter keeps track of number of vertices in taken set to true
        while (counter < n){ // while not all vertices are in MST
            double lowest = INF; int v = 0; //scan A to get v where A[v].first is minimum in A
            for (int j = 0; j < n; j++){
                if (A[j].first() < lowest){
                    lowest = A[j].first();
                    v = j;
                }
            }
            if (!taken[v]) {counter += 1; taken[v] = true;}
            if (!taken[A[v].second()]) {counter += 1; taken[A[v].second()] = true;}
            cost += A[v].first();
            A[v] = new DoubleIntPair(INF, A[v].second()); //prevent picking the same point
            for (int j = 0; j < n; j++){
                if (!taken[j] && A[j].first() > distance(points[v].x(), points[v].y(), points[j].x(), points[j].y()) - points[v].radius() - points[j].radius()){ //subtract radius of two circles from distance between their two centres
                    A[j] = new DoubleIntPair(distance(points[v].x(), points[v].y(), points[j].x(), points[j].y()) - points[v].radius() - points[j].radius(), v);
                }
            }
        }
        pw.println(cost);
        br.close();
        pw.close();
    }
}